scipy/optimize/tests/test_zeros.py

```diff
@@ -205,6 +206,333 @@ def test_lru_cached_individual(self, method):
+        res1 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res1.fun), 1e-3)
+        kwargs['fatol'] = 1e-6
+        res2 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res2.fun), 1e-6)
+        assert_array_less(np.abs(res2.fun), np.abs(res1.fun))
+
+        kwargs = kwargs0.copy()
+        kwargs['frtol'] = 1e-3
+        x1, x2 = bracket
+        f0 = np.minimum(abs(self.f(x1, *args)), abs(self.f(x2, *args)))
+        res1 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res1.fun), 1e-3*f0)
+        kwargs['frtol'] = 1e-6
+        res2 = zeros._chandrupatla(self.f, *bracket, **kwargs)
+        assert_array_less(np.abs(res2.fun), 1e-6*f0)
+        assert_array_less(np.abs(res2.fun), np.abs(res1.fun))
+
+    def test_maxiter_callback(self):
+        # Test behavior of `maxiter` parameter and `callback` interface
+        p = 0.612814
+        dist = stats.norm()
+        bracket = (-5, 5)
+        maxiter = 5
+
+        res = zeros._chandrupatla(self.f, *bracket, args=(dist, p),
+                                  maxiter=maxiter)
+        assert not np.any(res.success)
+        assert np.all(res.nfev == maxiter+2)
+        assert np.all(res.nit == maxiter)
+
+        def callback(res):
+            callback.iter += 1
+            callback.res = res
+            assert hasattr(res, 'x')
+            if callback.iter == 0:
+                # callback is called once with initial bracket
+                assert res.xl, res.xr == bracket
+            assert res.status == zeros._EINPROGRESS
+            if callback.iter == maxiter:
+                raise StopIteration
+        callback.iter = -1  # callback called once before first iteration
+        callback.res = None
+
+        res2 = zeros._chandrupatla(self.f, *bracket, args=(dist, p),
+                                   callback=callback)
+
+        # terminating with callback is identical to terminating due to maxiter
+        # (except for `status`)
+        for key in res.keys():
+            if key == 'status':
+                assert res[key] == zeros._ECONVERR
+                assert callback.res[key] == zeros._EINPROGRESS
+                assert res2[key] == zeros._ECALLBACK
+            else:
+                assert res2[key] == callback.res[key] == res[key]
+
+    @pytest.mark.parametrize('case', optimize._tstutils._CHANDRUPATLA_TESTS)
+    def test_nit_expected(self, case):
+        # Test that `_chandrupatla` implements Chandrupatla's algorithm:
+        # in all 40 test cases, the number of iterations performed
+        # matches the number reported in the original paper.
+        f, bracket, root, nfeval, id = case
+        # Chandrupatla's criterion is equivalent to
+        # abs(x2-x1) < 4*abs(xmin)*xrtol + xatol, but we use the more standard
+        # abs(x2-x1) < abs(xmin)*xrtol + xatol. Therefore, set xrtol to 4x
+        # that used by Chandrupatla in tests.
```

Comment:
I see, the explanation is here (also realising this is in the Notes section.) This should also be close to the code.

Reply:
I think I addressed this in https://github.com/scipy/scipy/pull/17719#discussion_r1234575150